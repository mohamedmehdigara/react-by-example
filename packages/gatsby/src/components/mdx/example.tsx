import React, { FC, ReactNode, useCallback, useState } from 'react'
import styled from 'styled-components'
import { CodeSandboxIcon, ResetIcon } from '../icons'

const Render = styled.div`
  display: flex;
  padding: 1em;
  border: 1px solid var(--app-foreground);
  border-radius: var(--app-border-radius);
  align-items: center;
  justify-content: center;
`
const FlexRight = styled.div`
  display: flex;
  justify-content: flex-end;
`

const CodeSandboxLink = styled.a`
  display: flex;
`

const ResetButton = styled.button`
  display: flex;
  margin: 0;
  padding: 0;
  color: var(--app-foreground);
  background: none;
  border: none;
  cursor: pointer;

  :focus,
  :active {
    color: var(--app-cyan);
    outline: 2px solid var(--app-cyan);
  }
`

export interface ExampleProps {
  code?: ReactNode
  render?: ReactNode
  link?: string
}

export const Example: FC<ExampleProps> = ({ code, render, link }) => {
  const [resetKey, setResetKey] = useState(Date.now())
  const triggerReset = useCallback(() => {
    setResetKey(Date.now())
  }, [])
  return (
    <>
      {code}
      <Render key={resetKey}>
        <div>{render}</div>
      </Render>
      {link && (
        <FlexRight>
          <ResetButton
            type="button"
            title="reset example"
            onClick={triggerReset}
          >
            <ResetIcon />
          </ResetButton>
          <CodeSandboxLink
            target="_blank"
            rel="noreferrer"
            href={link}
            title="edit on CodeSandbox"
          >
            <CodeSandboxIcon />
          </CodeSandboxLink>
        </FlexRight>
      )}
    </>
  )
}
