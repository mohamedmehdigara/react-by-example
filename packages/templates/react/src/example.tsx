import React, { FC, ReactElement } from 'react'

export const Simple: FC = () => <h1>Hello world!</h1>

export default (): ReactElement => <Simple />
