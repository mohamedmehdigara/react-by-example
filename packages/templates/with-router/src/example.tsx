import React, { FC } from 'react'

export const Simple: FC = () => <h1>Hello world!</h1>

export default <Simple />
