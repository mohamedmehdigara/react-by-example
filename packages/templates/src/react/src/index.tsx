import React from 'react'
import ReactDOM from 'react-dom'
import example from './example'

ReactDOM.render(<>{example}</>, document.getElementById('root'))
