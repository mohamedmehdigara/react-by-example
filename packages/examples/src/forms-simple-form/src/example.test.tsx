import { cleanup, render, RenderResult } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import React from 'react'
import { SimpleForm } from './example'
import { UserAPI } from './user-api'

const matchDebug = (partial: Record<string, unknown>): RegExp =>
  new RegExp(JSON.stringify(partial, null, '  ').slice(4, -2))

describe('SimpleForm', () => {
  let component: RenderResult
  let form: HTMLElement
  let username: HTMLElement
  let password: HTMLElement
  let terms: HTMLElement
  let submit: HTMLElement
  let debug: HTMLElement
  let registerSpy: jest.SpyInstance

  beforeEach(() => {
    component = render(<SimpleForm />)
    form = component.getByRole('form')
    username = component.getByLabelText('Username')
    password = component.getByLabelText('Password')
    terms = component.getByLabelText(/Terms/)
    submit = component.getByRole('button')
    debug = component.getByText(matchDebug({ username: '' }))
    registerSpy = jest.spyOn(UserAPI, 'register').mockRejectedValue(new Error())
  })

  afterEach(() => {
    cleanup()
  })

  it('should snapshot', () => {
    expect(component.container).toMatchSnapshot()
  })

  describe('when entering username', () => {
    beforeEach(async () => {
      await userEvent.type(username, 'my_username')
    })

    it('should update input', () => {
      expect(username).toHaveValue('my_username')
    })

    it('should update debug', () => {
      expect(debug).toHaveTextContent(matchDebug({ username: 'my_username' }))
    })

    describe('when entering password', () => {
      beforeEach(async () => {
        await userEvent.type(password, '12345678')
      })

      it('should update input', () => {
        expect(password).toHaveValue('12345678')
      })

      it('should update debug', () => {
        expect(debug).toHaveTextContent(matchDebug({ password: '12345678' }))
      })

      describe('when accepting terms', () => {
        beforeEach(() => {
          userEvent.click(terms)
        })

        it('should update input', () => {
          expect(terms).toBeChecked()
        })

        it('should update debug', () => {
          expect(debug).toHaveTextContent(matchDebug({ terms: true }))
        })

        it('should have all form values', () => {
          expect(form).toHaveFormValues({
            username: 'my_username',
            password: '12345678',
            terms: true,
          })
        })

        describe('when submitting', () => {
          let event: Event

          beforeEach(() => {
            document.addEventListener(
              'submit',
              (e) => {
                event = e
              },
              { once: true },
            )

            userEvent.click(submit)
          })

          it('should prevent default', () => {
            expect(event).toMatchObject({ defaultPrevented: true })
          })

          it('should update debug', () => {
            expect(debug).toHaveTextContent(matchDebug({ submitted: true }))
          })

          it('should register', () => {
            expect(registerSpy).toHaveBeenCalledWith({
              username: 'my_username',
              password: '12345678',
            })
          })
        })
      })
    })
  })
})
