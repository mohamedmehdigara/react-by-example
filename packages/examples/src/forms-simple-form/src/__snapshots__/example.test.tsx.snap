// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SimpleForm should snapshot 1`] = `
<div>
  <form
    aria-label="registration form"
  >
    <div>
      <label>
        Username
        <input
          name="username"
          type="text"
          value=""
        />
      </label>
    </div>
    <div>
      <label>
        Password
        <input
          name="password"
          type="password"
          value=""
        />
      </label>
    </div>
    <div>
      <label>
        <input
          name="terms"
          type="checkbox"
        />
        <span>
          I have read the Terms and Conditions
        </span>
      </label>
    </div>
    <button
      type="submit"
    >
      Register
    </button>
    <pre>
      {
  "username": "",
  "password": "",
  "terms": false,
  "submitted": false
}
    </pre>
  </form>
</div>
`;
